railties/lib/rails/generators/rails/app/app_generator.rb

```diff
@@ -244,11 +251,31 @@ def create_test_files
       end
 
       def create_tmp_files
-        build(:tmp)
+        build(:tmp) unless options[:api]
       end
 
       def create_vendor_files
-        build(:vendor)
+        build(:vendor) unless options[:api]
```

Comment:
I think vendor is also needed in an API application. For example, I usually store internal gems engines inside `vendor/gems` to keep them in the same project.